core/connection/manager.go

```diff
@@ -240,25 +235,35 @@ func (manager *connectionManager) startConnection(consumerID identity.Identity,
 	}
 
 	go manager.consumeConnectionStates(stateChannel)
-	go connectionWaiter(connection, dialog)
+	go manager.connectionWaiter(connection)
 	return nil
 }
 
 func (manager *connectionManager) Status() Status {
-	manager.mutex.RLock()
-	defer manager.mutex.RUnlock()
+	manager.statusLock.RLock()
+	defer manager.statusLock.RUnlock()
 
 	return manager.status
 }
 
+func (manager *connectionManager) setStatus(cs Status) {
+	manager.statusLock.Lock()
+	manager.status = cs
+	manager.statusLock.Unlock()
+}
+
 func (manager *connectionManager) Disconnect() error {
-	manager.mutex.RLock()
-	defer manager.mutex.RUnlock()
+	manager.discoLock.Lock()
```

Comment:
Do we really need a second mutex for `Disconnect`? 
We already have a mutex for status, we can check status here to make sure we are in a right state.

Reply:
I think it's required here due to the fact that the status is a r/w lock. We could get into a situation where 2 goroutines are calling disconnect here. What could happen then is:
1st routine checks the status state, let's say it's connected. So it get's that. It's execution get's delayed at this point in time.
2nd routine checks the status - we're still connected, alrighty.

Both  1st and 2nd routines start executing the cleanConnection, which I want to avoid.

The status though should definitely be behind a r/wlock since it needs to be more responsive for external calls to check the status.